src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2100,6 +2132,50 @@ private SafeContext GetInvocationEscapeWithUpdatedRules(
             return escapeScope;
         }
 
+        private static void ReplaceWithExtensionImplementationIfNeeded(ref MethodInfo methodInfo, ref ImmutableArray<ParameterSymbol> parameters,
+            ref BoundExpression? receiver, ref ImmutableArray<BoundExpression> argsOpt, ref ImmutableArray<RefKind> argRefKindsOpt,
+            ref ImmutableArray<int> argsToParamsOpt)
+        {
+            Symbol? symbol = methodInfo.Symbol;
+            if (symbol?.GetIsNewExtensionMember() != true || symbol.IsStatic)
+            {
+                return;
+            }
+
+            MethodInfo replacedMethodInfo = methodInfo.ReplaceWithExtensionImplementation(out bool wasError);
+            if (wasError)
+            {
+                return;
+            }
+
+            methodInfo = replacedMethodInfo;
+
+            var extensionParameter = symbol.ContainingType.ExtensionParameter;
+            Debug.Assert(extensionParameter is not null);
+            parameters = parameters.IsDefault ? [extensionParameter] : [extensionParameter, .. parameters];
+
+            Debug.Assert(receiver is not null);
+            argsOpt = argsOpt.IsDefault ? [receiver] : [receiver, .. argsOpt];
+            receiver = null;
+
+            if (!argRefKindsOpt.IsDefault)
+            {
+                argRefKindsOpt = [RefKind.None, .. argRefKindsOpt];
```

Comment:
Why RefKind.None and not extensionParameter.RefKind?

Reply:
Because the argument is passed without an explicit ref kind, just like the receiver of a classic extension method.